cpp/tests/io/parquet_test.cpp

```diff
@@ -1048,13 +1048,21 @@ class custom_test_data_sink : public cudf::io::data_sink {
   bool supports_device_write() const override { return true; }
 
   void device_write(void const* gpu_data, size_t size, rmm::cuda_stream_view stream) override
+  {
+    this->device_write_async(gpu_data, size, stream).get();
+  }
+
+  std::future<void> device_write_async(void const* gpu_data,
+                                       size_t size,
+                                       rmm::cuda_stream_view stream) override
   {
     char* ptr = nullptr;
     CUDA_TRY(cudaMallocHost(&ptr, size));
     CUDA_TRY(cudaMemcpyAsync(ptr, gpu_data, size, cudaMemcpyDeviceToHost, stream.value()));
     stream.synchronize();
     outfile_.write(ptr, size);
     CUDA_TRY(cudaFreeHost(ptr));
+    return std::async(std::launch::deferred, [] {});
```

Comment:
Should this be calling `stream.synchronize` in the `device_write_async` method or only in the sync method? When using the async method, would it be possible to defer stream sync until after all threads have executed? Trying to make sure I understand the execution model correctly here. I'm doing some reading on this topic now.

Reply:
I don't think this test means anything other than to test the interface while making a custom sink. It still uses host_write in it. `device_write_async` was added because it's a mandatory part of that interface now.

Reply:
I could change this to the following to bring some asynchronicity to it
```c++
    return std::async(std::launch::deferred, [=] {
      char* ptr = nullptr;
      CUDA_TRY(cudaMallocHost(&ptr, size));
      CUDA_TRY(cudaMemcpyAsync(ptr, gpu_data, size, cudaMemcpyDeviceToHost, stream.value()));
      stream.synchronize();
      outfile_.write(ptr, size);
      CUDA_TRY(cudaFreeHost(ptr));
    });
```